### DISCUSSION TITLE
Tool to detect unused locale keys

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#7978] [Proposals]
**Describe the problem you would like to solve**
As the code evolves some locale keys might become outdated/unused, and removing them would probably make the life of translators easier, especially for new languages/scripts.

**Describe the solution you'd like**
Probably a tool to be executed at building (if it's fast enough) or developing time, to display a summary, such as `en_US: 2 possible unused locale keys were found: (...)` would be useful.

Checks:
- Any locale key that doesn't exist in the main locale (en_US)
- Locale keys that cannot be found in the source code
- Duplicated locale keys (present in more than one locale file)
- Perhaps duplicated translations can be merged into a single locale key

Notes:
- Check for an easy way (ready code) to skip .gitignored files
- Application specific locale keys, that are present and have the same translations across all applications (e.g. ops/ojs/omp), could be moved to the `pkp-lib`
- We already have tools to remove/migrate the locale keys, which could be updated to use the `gettext` package

**Who is asking for this feature?**
Developer

**Comment by asmecher** [comment]
@bozana asked about this a few weeks back -- we used to have a tool when we were using XML-based locale keys. It would need a rewrite, but it's here: https://github.com/pkp/ojs/commit/baf1ce4d393e220cd7c792a353129ddfa3d8e0d9

**Comment by NateWr** [comment]
> Locale keys that cannot be found in the source code

One thing to be wary of is that there are places where we are concatenating the locale keys. Things like this:

```php
foreach ($types as $type) {
    $name = __('some.example.' . $type);
}
```

I've been trying over the years to remove such things when I find them. And I think we should avoid this in any future code. But we'll need to be careful.
